### PR TITLE
fix: BalanceInput

### LINKED ISSUE #3520: "Make offer" - dont recalculate in another unit of measure
### What happened?

Wallet:  bXiGvoYjMbEEUe27h3zFCnabsu367zEDEivu8ZHjVSm5v8mC2

Hello again! :)

Another one mistake: 
When I make offer - it controls only 10000. But there is no control for measures.  
For example, 10000 BSX its not the same as 10000 nanoBSX.
And then I get an Error. 

Example for reproducing in the video:

https://user-images.githubusercontent.com/96058877/180233595-77826434-7450-4736-9de0-948049c0e302.mp4

  

### Please reproduce in steps

All the steps in the video:

Uploading bandicam 2022-07-21 16-58-22-558.mp4…

### Expected Behavior

Recalculating limit to new measure when the unit of measure is changed

### What browsers are you seeing the problem on?

Chrome

### At which address did you encounter bug?

beta.kodadot.xyz

### Are you logged in?

Yes

### Which wallet you are using?

PolkadotJS

### Screenshots

https://user-images.githubusercontent.com/96058877/180234401-342dd108-9083-4ac9-b103-d210a837cb10.mp4

### Relevant log output

_No response_

### Payment link for reward

https://kodadot.xyz/transfer/?target=%3CMy_Kusama_Address_check_https://github.com/kodadot/nft-gallery/blob/main/CONTRIBUTING.md#creating-your-ksm-address%3E

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #3558: Balance Input emit wrong value after unit change
### What happened?

As title suggested.the 

### Please reproduce in steps

Any Gallery Item could reproduce this.

### Expected Behavior

BSX should be the base unit, and emit the right value with bsx unit after the value changed or unit changed.

### What browsers are you seeing the problem on?

_No response_

### At which address did you encounter bug?

_No response_

### Are you logged in?

_No response_

### Which wallet you are using?

_No response_

### At which chain did you encounter bug?

_No response_

### Screenshots

image

### Relevant log output

_No response_

### Payment link for reward

https://kodadot.xyz/transfer/?target=%3CMy_Kusama_Address_check_https://github.com/kodadot/nft-gallery/blob/main/CONTRIBUTING.md#creating-your-ksm-address%3E

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by KngZhi**
- feat: alert info with right value with different unit
- feat: emit value in default unit(1)
- fix: unit change
- fix: disable submit while empty input

**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [x] Bugfix
- [ ] Feature
- [x] Refactoring

### What's new?

- [x] PR closes #3558 
- [x] PR closes #3520 
- [ ] Requires deployment <rubick/magick_issue>
- [ ] <brief_description_of_what_I've_added>

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

### Optional

- [ ] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [ ] I've tested PR on mobile and everything seems works
- [ ] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer?target=DaWoVNNhBTdqJ5iPN3UcaLeamkPpJeGWdrs8LhvX9TQbadc)

### Community participation

- [ ] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

https://user-images.githubusercontent.com/11361147/181010982-3b1526cf-d351-4ecc-b902-4ca6f5a134c4.mov

### CODE REVIEW

**Comment by KngZhi** [comment]
@kkukelka @petersopko may I also bother you to review this PR as well, many thanks :)

I saw the git history, you are the main contributor for this component.

File: components/bsx/Gallery/Item/AvailableActions.vue [hunk 1]
```diff
@@ -159,9 +159,12 @@ export default class AvailableActions extends mixins(
 
   protected updateMeta(value: string | number) {
     const balanceInputComponent = this.$refs.balanceInput as BalanceInput
-    this.isBalanceInputValid = balanceInputComponent.checkValidity()
-    this.$consola.log(typeof value, value)
     this.meta = value
+    if (this.meta === 0 || this.meta === '0') {
```
Comment (Jarsen136) [inline components/bsx/Gallery/Item/AvailableActions.vue:163]: there is something duplicate

**Comment by vikiival** [comment]
Love to have some tests for this one

0d416015224ee8bcf7d20b8b706eb809pepe-holding-a-cross

**Comment by KngZhi** [comment]
> Love to have some tests for this one

Made a thorough test by myself.

For the component unit test, maybe we can create a new issue for this since we have to add more packages to do that and solve some conflicts. Tried before, but there are some issues with *Buefy* that vue-util can not select b-component to do testing.

image

**Comment by vikiival** [comment]
> For the component unit test, maybe we can create a new issue for this since we have to add more packages to do that and solve some conflicts. Tried before, but there are some issues with Buefy that vue-test-utils can not find b-component to do testing.

Sure!

**Comment by JustLuuuu** [comment]
Im unable to add new NFT to my collection: Impossible to click on "Click to create NFT(s)", not sure why
image

**Comment by JustLuuuu** [comment]
Unable to add new NTF to my collection. Not possible to click on "Click on to create NFT(s), not sure why
image

File: components/shared/BalanceInput.vue [hunk 2]
```diff
@@ -88,16 +81,18 @@ export default class BalanceInput extends mixins(ChainMixin) {
...
-  @Emit('input')
   public handleInput(value: number) {
     this.internalValue = value
+    this.$emit('input', String(this.internalValue * this.selectedUnit))
     return String(value)
   }
 
   handleUnitChange(unit) {
-    this.selectedUnit = unit
+    const valueInBSXUnit = this.internalValue * this.selectedUnit
```
Comment (vikiival) [inline components/shared/BalanceInput.vue:91]: dafuq

Comment (KngZhi) [reply components/shared/BalanceInput.vue:91]: unit exchange.

e.g., input as 10000 BSX <-> 10 Kilo while unit change. Firstly convert all units into BSX unit first and then convert to the selected unit.

**Comment by KngZhi** [comment]
> Unable to add new NTF to my collection. Not possible to click on "Click on to create NFT(s), not sure why image

the beta site can not work as well, this may be caused by some merges. Since previously I remember it worked.

image

File: components/shared/BalanceInput.vue (hunk 2 above)
Comment (vikiival) [reply components/shared/BalanceInput.vue:91]: I am not sure if BSX sounds like share imo

Screenshot 2022-07-27 at 20 13 17

Comment (KngZhi) [reply components/shared/BalanceInput.vue]: changed to valueInBaseUnit.

Comment (vikiival) [reply components/shared/BalanceInput.vue]: thanks ^-^

**Review by JustLuuuu — APPROVED** [review]
wfm

**Comment by yangwao** [comment]
@KngZhi recently balanceInput.vue has changes in #3595 

let's resolve them and let's go

**Comment by KngZhi** [comment]
okay, outside now, would fix that asap.

**Comment by KngZhi** [comment]
hi @yangwao solved the conflict, is this okay to merge?

File: components/bsx/Gallery/Item/AvailableActions.vue [hunk 3]
```diff
@@ -183,6 +183,12 @@ export default class AvailableActions extends mixins(
     }
     this.$consola.log(typeof value, value)
     this.meta = value
+    // ad-hoc fix for empty input value
+    if (this.meta === '0') {
+      this.isBalanceInputValid = false
+    } else {
+      this.isBalanceInputValid = balanceInputComponent.checkValidity()
+    }
```
Comment (Jarsen136) [inline components/bsx/Gallery/Item/AvailableActions.vue:191]: Before calling `checkValidity`, we should make sure `balanceInputComponent instanceof BalanceInput`.

Or there will be some error.

**Review by Jarsen136 — COMMENTED** [review]
Gifting out nft
image

File: components/bsx/Gallery/Item/AvailableActions.vue (hunk 3 above)
Comment (KngZhi) [reply components/bsx/Gallery/Item/AvailableActions.vue:187]: made a duplicate check, and removed an unnecessary one.
Also, add a guard for BlanceInput value check.

btw, that's why viki mentioned we should avoid `ref`...

File: components/bsx/Gallery/Item/AvailableActions.vue [hunk 4]
```diff
@@ -180,6 +180,14 @@ export default class AvailableActions extends mixins(
...
       this.isBalanceInputValid = balanceInputComponent.checkValidity()
+      // ad-hoc fix for empty input value
+      if (this.meta === '0') {
+        this.isBalanceInputValid = false
+      }
+    }
+    // ad-hoc fix for empty input value
+    if (balanceInputComponent instanceof BalanceInput && this.meta === '0') {
+      this.isBalanceInputValid = false
     }
```
Comment (Jarsen136) [inline components/bsx/Gallery/Item/AvailableActions.vue:191]: pls delete these duplicate code : )

Comment (KngZhi) [reply components/bsx/Gallery/Item/AvailableActions.vue]: using `git amend` to cause that, sorry for that. removed.

**Review by Jarsen136 — APPROVED** [review]
LGTM

**Comment by Jarsen136** [comment]
pay 100 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $100 @ 66.09 USD/KSM ~ 1.513 $KSM
🧗 [DaWoVNNhBTdqJ5iPN3UcaLeamkPpJeGWdrs8LhvX9TQbadc](https://kusama.subscan.io/account/DaWoVNNhBTdqJ5iPN3UcaLeamkPpJeGWdrs8LhvX9TQbadc)
🔗 [0x9644165ab8206ecfdb567e204d5b08eaa7f3178c45b5fb6c4a8fb5f9e46d252d](https://kusama.subscan.io/extrinsic/0x9644165ab8206ecfdb567e204d5b08eaa7f3178c45b5fb6c4a8fb5f9e46d252d)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**